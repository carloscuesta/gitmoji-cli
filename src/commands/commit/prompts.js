// @flow
import inquirer from 'inquirer'
import inquirerAutocompletePrompt from 'inquirer-autocomplete-prompt'

import configurationVault from '@utils/configurationVault'
import filterGitmojis from '@utils/filterGitmojis'
import getDefaultCommitContent from '@utils/getDefaultCommitContent'
import { type CommitOptions, capitalizeTitle } from '.'
import guard from './guard'
import Fuse from 'fuse.js'

const TITLE_MAX_LENGTH_COUNT: number = 48

inquirer.registerPrompt('autocomplete', inquirerAutocompletePrompt)

export type Gitmoji = {
  code: string,
  description: string,
  emoji: string,
  name: string
}

export type Answers = {
  gitmoji: string,
  scope: ?string,
  title: string,
  message: ?string
}

export const filterScopes = (
  input?: string,
  scopes: Array<string>
): Function | Array<string> => {
  const fuse = new Fuse(scopes, {
    includeScore: true
  })

  return input ? fuse.search(input).map((scope) => scope.item) : scopes
}

export default (
  gitmojis: Array<Gitmoji>,
  options: CommitOptions
): Array<Object> => {
  const { title, message, scope } = getDefaultCommitContent(options)

  const getScopePrompt = (scopes?: boolean | string[]): Array<Object> => {
    const name = 'scope'
    const message = 'Enter the scope of current changes:'

    switch (scopes) {
      case false:
      case undefined:
        return []
      case true:
        return [
          {
            name,
            message,
            ...(scope ? { default: scope } : {})
          }
        ]
      default:
        return [
          {
            name,
            message,
            ...(scope ? { default: scope } : {}),
            type: 'autocomplete',
            source: (answersSoFor: any, input: string) => {
              return Promise.resolve(
                filterScopes(input, scopes).map((scopeSearch) => ({
                  name: scopeSearch,
                  value: scopeSearch
                }))
              )
            }
          }
        ]
    }
  }

  return [
    {
      name: 'gitmoji',
      message: 'Choose a gitmoji:',
      type: 'autocomplete',
      source: (answersSoFor: any, input: string) => {
        return Promise.resolve(
          filterGitmojis(input, gitmojis).map((gitmoji) => ({
            name: `${gitmoji.emoji}  - ${gitmoji.description}`,
            value: gitmoji[configurationVault.getEmojiFormat()]
          }))
        )
      }
    },
    ...getScopePrompt(configurationVault.getScopePrompt()),
    {
      name: 'title',
      message: 'Enter the commit title',
      validate: guard.title,
      transformer: (input: string) => {
        const length = (title || input).length.toString().padStart(2, '0')

        return `[${length}/${TITLE_MAX_LENGTH_COUNT}]: ${
          configurationVault.getCapitalizeTitle()
            ? capitalizeTitle(input)
            : input
        }`
      },
      ...(title ? { default: title } : {})
    },
    ...(configurationVault.getMessagePrompt()
      ? [
          {
            name: 'message',
            message: 'Enter the commit message:',
            ...(message ? { default: message } : {})
          }
        ]
      : [])
  ]
}
