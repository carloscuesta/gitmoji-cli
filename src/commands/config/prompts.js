// @flow
import configurationVault from '../../utils/configurationVault'

export const CONFIGURATION_PROMPT_NAMES = {
  AUTO_ADD: 'autoAdd',
  EMOJI_FORMAT: 'emojiFormat',
  SCOPE_PROMPT: 'scopePrompt',
  SIGNED_COMMIT: 'signedCommit'
}

export const EMOJI_COMMIT_FORMATS = {
  CODE: 'code',
  EMOJI: 'emoji'
}

export default () => [
  {
    name: CONFIGURATION_PROMPT_NAMES.AUTO_ADD,
    message: 'Enable automatic "git add ."',
    type: 'confirm',
    default: configurationVault.getAutoAdd()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.EMOJI_FORMAT,
    message: 'Select how emojis should be used in commits',
    type: 'list',
    choices: [
      { name: ':smile:', value: EMOJI_COMMIT_FORMATS.CODE },
      { name: 'ðŸ˜„', value: EMOJI_COMMIT_FORMATS.EMOJI }
    ],
    default: configurationVault.getEmojiFormat()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.SIGNED_COMMIT,
    message: 'Enable signed commits',
    type: 'confirm',
    default: configurationVault.getSignedCommit()
  },
  {
    name: CONFIGURATION_PROMPT_NAMES.SCOPE_PROMPT,
    message: 'Enable scope prompt',
    type: 'confirm',
    default: configurationVault.getScopePrompt()
  }
]
