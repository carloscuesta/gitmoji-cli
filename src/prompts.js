const constants = require('./constants')
const configVault = require('./config')
const guard = require('./guard')
const utils = require('./utils')

const config = [
  {
    name: constants.AUTO_ADD,
    message: 'Enable automatic "git add ."',
    type: 'confirm'
  },
  {
    name: constants.ISSUE_FORMAT,
    message: 'Choose Issue Format',
    type: 'list',
    choices: ['github', 'jira']
  },
  {
    name: constants.EMOJI_FORMAT,
    message: 'Select how emojis should be used in commits',
    type: 'list',
    choices: [
      { name: ':smile:', value: 'code' }, { name: 'ðŸ˜„', value: 'emoji' }
    ]
  },
  {
    name: constants.SIGNED_COMMIT,
    message: 'Enable signed commits',
    type: 'confirm'
  }
]

const gitmoji = (gitmojis) => {
  return [
    {
      name: 'gitmoji',
      message: 'Choose a gitmoji:',
      type: 'autocomplete',
      source: (answersSoFor, input) => {
        return Promise.resolve(
          gitmojis.filter((gitmoji) => {
            const emoji = gitmoji.name.concat(gitmoji.description).toLowerCase()
            return (!input || emoji.indexOf(input.toLowerCase()) !== -1)
          })
            .map((gitmoji) => ({
              name: `${gitmoji.emoji}  - ${gitmoji.description}`,
              value: gitmoji[configVault.getEmojiFormat() || constants.CODE]
            }))
        )
      }
    },
    {
      name: 'title',
      message: 'Enter the commit title',
      validate: guard.title,
      transformer: (input) => utils.inputCountTransformer(
        input,
        constants.TITLE_MAX_LENGTH_COUNT
      )
    },
    {
      name: 'message',
      message: 'Enter the commit message:',
      validate: guard.message
    },
    {
      name: 'reference',
      message: 'Issue / PR reference:',
      validate: (value) => guard.reference(value, configVault.getIssueFormat())
    }
  ]
}

module.exports = {
  config,
  gitmoji
}
