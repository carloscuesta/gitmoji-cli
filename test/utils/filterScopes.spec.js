import { filterScopes } from '../../src/commands/commit/prompts.js'
import * as stubs from './stubs'

describe('filterScopes', () => {
  it('should find all scopes with empty input', () => {
    const filteredGitmojis = filterScopes(undefined, stubs.scopes)

    expect(filteredGitmojis).toStrictEqual(stubs.scopes)
  })

  it('should find the `apps` scope first', () => {
    const filteredGitmojis = filterScopes('apps', stubs.scopes)

    const scope = stubs.scopes.find((scope) => scope === 'apps')
    expect(filteredGitmojis[0]).toStrictEqual(scope)
  })

  it('should find the `toolings/typescript` scope first', () => {
    const filteredGitmojis = filterScopes('toolings/types', stubs.scopes)

    const scope = stubs.scopes.find((scope) => scope === 'toolings/typescript')
    expect(filteredGitmojis[0]).toStrictEqual(scope)
  })

  it('should match a list of filteredScopes', () => {
    const filteredScopes = filterScopes('packages', stubs.scopes)

    expect(filteredScopes).toMatchSnapshot()
  })
})
