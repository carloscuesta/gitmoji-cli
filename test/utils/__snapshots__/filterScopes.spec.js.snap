// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterScopes should match a list of filteredScopes 1`] = `
[
  "packages",
  "packages/ui",
  "packages/shared",
]
`;
